TWTValidation/JSON Validator/Parser/TWTJSONSchemaParser.m

```diff
@@ -0,0 +1,652 @@
+{
+    [self checkObject:self.JSONSchema[kTWTJSONSchemaKeywordSchema] isOneOf:[NSSet setWithObject:kTWTJSONSchemaKeywordDraft4Path]];
+
+    TWTJSONSchemaTopLevelASTNode *topLevelNode = [[TWTJSONSchemaTopLevelASTNode alloc] init];
+    topLevelNode.schemaPath = self.JSONSchema[kTWTJSONSchemaKeywordSchema];
+
+    @try {
+        topLevelNode.schema = [self parseSchema:self.JSONSchema];
+    } @catch (NSException *exception) {
+        NSError *outError = exception.userInfo[kTWTJSONExceptionErrorKey];
+        if (errorPointer) {
+            *errorPointer = outError;
+            return nil;
+        }
+    }
+
+    return topLevelNode;
+}
+
+
+# pragma mark - Schema parser methods
+// starting point for all schema nodes (except top level)
+// directs the schema toward the appropriate parser method based on the value of "type"
+// type-specific parser methods are responsible for the entire process of creating a node from a schema, including:
+// 1) initializing the appropriate ASTNode 2) setting the type (if its not built into that ASTNode class) 3) parsing all general and type-specific properties 4) returning a completed node
+- (TWTJSONSchemaASTNode *)parseSchema:(NSDictionary *)schema
+{
+    NSParameterAssert(schema);
+
+    [self checkObject:schema isOfClass:[NSDictionary class] allowNil:NO];
+
+    if ([schema[kTWTJSONSchemaKeywordType] isKindOfClass:[NSString class]]) {
+        NSString *instanceType = schema[kTWTJSONSchemaKeywordType];
+
+        static NSDictionary *typeSelectors = nil;
+        static dispatch_once_t onceToken;
+        dispatch_once(&onceToken, ^{
+            typeSelectors = @{ kTWTJSONSchemaTypeKeywordAny : @"parseGenericTypeSchema:",
+                               kTWTJSONSchemaTypeKeywordArray : @"parseArrayTypeSchema:",
+                               kTWTJSONSchemaTypeKeywordBoolean : @"parseGenericTypeSchema:",
+                               kTWTJSONSchemaTypeKeywordInteger : @"parseNumberTypeSchema:",
+                               kTWTJSONSchemaTypeKeywordNull : @"parseGenericTypeSchema:",
+                               kTWTJSONSchemaTypeKeywordNumber : @"parseNumberTypeSchema:",
+                               kTWTJSONSchemaTypeKeywordObject : @"parseObjectTypeSchema:",
+                               kTWTJSONSchemaTypeKeywordString : @"parseStringTypeSchema:" };
+        });
+
+        if (typeSelectors[instanceType]) {
+            SEL parseSelector = NSSelectorFromString(typeSelectors[instanceType]);
+            NSAssert([self respondsToSelector:parseSelector], @"selector is not defined");
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
+            return [self performSelector:parseSelector withObject:schema];
+#pragma clang diagnostic pop
+        }
+    }
+
+    return [self parseGenericTypeSchema:schema];
+}
+
+
+- (TWTJSONSchemaArrayASTNode *)parseArrayTypeSchema:(NSDictionary *)arraySchema
```

Comment:
In all of these, I think we can drop the word “type” from the method name. `parseArraySchema:`.